src/aspire/abinitio/commonline_c3_c4.py

```diff
@@ -231,21 +237,270 @@ def _estimate_inplane_rotations(
     # Secondary Methods for computing outer product #
     #################################################
 
-    def _self_clmatrix_c3_c4(self, pf, order, max_shift, shift_step):
-        # return sclmatrix
-        pass
+    def _self_clmatrix_c3_c4(self):
+        """
+        Find the single pair of self-common-lines in each image assuming that the underlying
+        symmetry is C3 or C4.
+        """
+        pf = self.pf.copy()
+        n_img = self.n_img
+        n_theta = self.n_theta
+        max_shift_1d = np.ceil(2 * np.sqrt(2) * self.max_shift)
+        shift_step = self.shift_step
+        order = self.order
 
-    def _estimate_all_Riis_c3_c4(order, sclmatrix, n_theta):
-        # return Riis
-        pass
+        # The angle between self-common-lines is in the range [60, 180] for C3 symmetry
+        # and [90, 180] for C4 symmetry. Since antipodal lines are perfectly correlated
+        # we search for common lines in a smaller window.
+        # Note: matlab code used [60, 165] for order=3 and [90, 160] for order=4.
+        # We've increased this window to improve the MSE of our estimates.
+        if order == 3:
+            min_angle_diff = 60 * np.pi / 180
+            max_angle_diff = 175 * np.pi / 180
+        else:
+            min_angle_diff = 90 * np.pi / 180
+            max_angle_diff = 175 * np.pi / 180
+
+        # The self-common-lines matrix holds two indices per image that represent
+        # the two self common-lines in the image.
+        sclmatrix = np.zeros((n_img, 2))
+        corrs_stats = np.zeros(n_img)
+        shifts_stats = np.zeros(n_img)
+
+        # We create a mask associated with angle differences that fall in the
+        # range [min_angle_diff, max_angle_diff].
+        X, Y = np.meshgrid(range(n_theta), range(n_theta // 2))
+        diff = Y - X
+        unsigned_angle_diff = np.arccos(np.cos(diff * 2 * np.pi / n_theta))
+        good_diffs = np.logical_and(
+            min_angle_diff < unsigned_angle_diff, unsigned_angle_diff < max_angle_diff
+        )
 
-    def _estimate_all_Rijs_c3_c4(order, clmatrix, n_theta):
-        # return Rijs
-        pass
+        # Compute the correlation over all shifts.
+        # Generate Shifts.
+        r_max = pf.shape[0]
+        shifts, shift_phases, _ = self._generate_shift_phase_and_filter(
+            r_max, max_shift_1d, shift_step
+        )
+        n_shifts = len(shifts)
+        all_shift_phases = shift_phases.T
 
-    def local_sync_J_c3_c4(order, Rijs, Riis):
-        # return vijs, viis
-        pass
+        # Transpose pf and reconstruct the full polar Fourier for use in correlation.
+        # self.pf only consists of rays in the range [180, 360).
+        pf = pf.transpose((2, 1, 0))
```

Comment:
The original axes are column major: (n_rad -1, n_theta//2, n_img), which comes from `_build()` in the base class `CLOrient3D`. This is something we might considering changing, since it is transposed when it is used for `CLSyncVoting` correlations as well.

I guess I could just use `numpy.T` here.

Reply:
Ok let's add a comment explaining that and start an issue.

Reply:
Ok, added comment and created issue #662.